webview/wsgi.py

```diff
@@ -0,0 +1,423 @@
+            rv += str(total)
+        else:
+            rv += '*'
+        return rv
+
+    def _serve_partial_file(self, environ, start_response, file, filename, mime):
+        response_headers = self._default_headers(mime, file)
+        length = response_headers['Content-Length']
+        if length:
+            length = int(length)
+        else:
+            length = None
+        start, end = self._parse_range(environ['HTTP_RANGE'], length)
+
+        if length is not None:
+            # Check ranges
+            maxindex = length - 1
+            if start > maxindex or end > maxindex:
+                start_response('416 Range Not Satisfiable', [
+                    ('Content-Range', 'bytes */{}'.format(length))
+                ])
+                return []
+
+        assert start <= end
+        assert length is None or end < length
+
+        logger.debug("Serving %s (%s to %s of %s)", filename, start, end, length)
+
+        response_headers['Content-Range'] = self._compose_content_range(start, end, length)
+        if end is None:
+            amount = None
+            del response_headers['Content-Length']
+        else:
+            amount = end - start + 1
+            response_headers['Content-Length'] = str(amount)
+
+        start_response('206 Partial Content', response_headers._headers)
+
+        if environ['REQUEST_METHOD'] == 'HEAD':
+            file.close()
+            return []
+        else:
+            return self._partial_file_wrapper(file, start, amount)
+
+    def _partial_file_wrapper(self, file, skip, amount):
+        served = 0
+
+        if skip:
+            file.seek(skip)
+
+        while (amount is None) or (served <= amount):
+            data = file.read(min(CHUNK_SIZE, amount - served))
+            if not data:
+                break
+            served += len(data)
+            yield data
+
+        logging.debug("Served %s of %s", served, amount)
+
+
+class StaticFiles(StaticContentsApp):
+    """
+    Serves static files from a directory on the file system.
+    """
+    def __init__(self, root):
+        self.root = abspath(root)
+
+    def open(self, file):
+        if file:
+            path = os.path.join(self.root, file.lstrip('/'))
+        else:
+            path = self.root
+        logger.debug("Resolved %s to %s", file, path)
+        return open(path, 'rb')
+
+
+class StaticResources(StaticContentsApp):
+    """
+    Serves static files from resources in python packages
+    """
+    def __init__(self, root):
+        self.root = root
+
+    def open(self, file):
+        slashed, basename = posixpath.split(file)
+        slashed = slashed.rstrip('/')
+        if slashed:
+            packagename = "{}.{}".format(self.root, slashed.replace('/', '.'))
+        else:
+            packagename = self.root
+        try:
+            return importlib_resources.open_binary(packagename, basename)
+        except ModuleNotFoundError:
+            raise FileNotFoundError
```

Comment:
`wsgi.py` and `serving.py` could be moved to the `http` subdirectory.

Reply:
Yeah, thought about that, not sure it would help anything? The names seem descriptive and there's not really any clutter.

Reply:
It is a personal preference, but I am fine either way. If this implementation grows beyond two files, then it would warrant an own directory.